### PR TITLE
added specs documentation

### PR DESCRIPTION
**Pull request by mgc1194**
This PR adds documentation on how to write unit tests in ruby to our style guide.

## Motivation
Specs syntax was introduced to the Platform team a couple of years ago. Since then, we have adopted this behaviour-driven development style facilitating the communication and understanding among team members about what the code is supposed to do. We have been developing tests in a way that describes the expected behaviour of the system to reflect product needs and user experience. 

We see the benefits of extending the use of the Spec syntax to the larger engineering team. Especially given the large surface area and complexity of our codebase, and the fact that we all work on a mono-repo that requires multiple teams touching the same parts of the code.

### Readability
- **Unit Test Style**: More structured and familiar to those with experience in other xUnit frameworks. Method names and class structures clarify what is being tested but may be less readable for those unfamiliar with the conventions.
- **Spec Style**: Emphasizes readability and natural language, making understanding what each test is doing easier, especially for those new to the codebase. This is especially important in a large team where engineers frequently read and modify tests written by others from different teams.

### Organization:
- **Unit Test Style**: Tests are organized by methods within classes. This can become unwieldy for large test suites with many tests.
- **Spec Style**: Tests are organized hierarchically using nested `describe` and `it` blocks, which can be more intuitive and easier to navigate. This can help manage the complexity of a large mono-repo by logically grouping related tests.

### Assertion Style:
- **Unit Test Style**: Uses assert/refute methods, which are straightforward but may require more boilerplate.
- **Spec Style**: Uses expectation-based assertions, which are more concise and often more expressive.

### Setup and Teardown:
- **Unit Test Style**: Uses `setup` and `teardown` methods, which apply to all tests within a class.
- **Spec Style**: Uses `before` and `after` blocks, which can be scoped to specific `describe` blocks, providing more granular control.

## Rollout

The choice between the two test styles will be up to personal or team preference, with some teams mixing both styles for different contexts.
- Use **Spec Style** if you want behaviour-driven development and want your tests to serve as readable documentation of the code’s behaviour.
- Use **Unit Test Style** if you are more comfortable with traditional xUnit frameworks and prefer a more structured approach.

## Links

- jira ticket: [P20-1127](https://codedotorg.atlassian.net/browse/P20-1127)
- [Better Specs](https://www.betterspecs.org/)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: STYLEGUIDE.md [hunk 1]
```diff
@@ -64,6 +64,62 @@ Fallback: https://github.com/styleguide/ruby
...
+
+      it 'returns another result' do
+        _my_method.must_equal 'expected result for argument_b'
+      end
+    end
+  end
+end
+```
+
+See a real example in our codebase in the `Queries::SectionTest` [test](https://github.com/code-dot-org/code-dot-org/blob/012a3bcb13f4a265d6570b8995ae872e92a0ab42/dashboard/test/lib/queries/section_test.rb) for the `Queries::Section` [Module](https://github.com/code-dot-org/code-dot-org/blob/012a3bcb13f4a265d6570b8995ae872e92a0ab42/dashboard/lib/queries/section.rb).
```
Comment (carl-codeorg) [inline STYLEGUIDE.md]: Another example is [omniauth_callbacks_controller_test.rb](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/controllers/omniauth_callbacks_controller_test.rb). It has some more complex examples. It also demonstrates how you can just add in spec-style tests even if there are other tests in the same class.

File: STYLEGUIDE.md [hunk 2]
```diff
@@ -64,6 +64,79 @@ Fallback: https://github.com/styleguide/ruby
     first_method.
     second_method
   ```
+  
+### Writing Unit Test
```
Comment (Hamms) [inline STYLEGUIDE.md]: super-tiny nit:
```suggestion
### Writing Unit Tests
```

File: STYLEGUIDE.md [hunk 3]
```diff
@@ -64,6 +64,79 @@ Fallback: https://github.com/styleguide/ruby
...
+Our unit tests testing library is [Minitest](https://github.com/minitest/minitest).
```
Comment (Hamms) [inline STYLEGUIDE.md]: nit:
```suggestion
Our unit testing library is [Minitest](https://github.com/minitest/minitest). 
```

File: STYLEGUIDE.md [hunk 4]
```diff
@@ -64,6 +64,79 @@ Fallback: https://github.com/styleguide/ruby
     first_method.
     second_method
   ```
+  
+### Writing Unit Test
+Our unit tests testing library is [Minitest](https://github.com/minitest/minitest). 
+
+When writing unit test we recommend to use [Specs Syntax](https://github.com/minitest/minitest?tab=readme-ov-file#specs-). Our default MiniTest Spec has been extended to include some RSpec features such as [`subject`](https://rspec.info/features/3-13/rspec-core/subject/explicit-subject/) or [`let!`](https://rspec.info/features/3-13/rspec-core/helper-methods/let/)
```
Comment (Hamms) [inline STYLEGUIDE.md]: nit:
```suggestion
When writing unit tests, we recommend you use the [Specs Syntax](https://github.com/minitest/minitest?tab=readme-ov-file#specs-). Our default MiniTest Spec has been extended to include some RSpec features such as [`subject`](https://rspec.info/features/3-13/rspec-core/subject/explicit-subject/) or [`let!`](https://rspec.info/features/3-13/rspec-core/helper-methods/let/)
```

File: STYLEGUIDE.md [hunk 5]
```diff
@@ -64,6 +64,79 @@ Fallback: https://github.com/styleguide/ruby
...
+
+This documentation represents a collection of references and resources for following best practices of Specs Syntax. It is intended to guide developers in writing clean, maintainable, and consistent unit tests using the Specs Syntax.
+
+However, it is important to acknowledge that much of our codebase (the majority of our codebase) does not actually follow this recommendation. This document serves as a guide to encourage moving towards these best practices in new code and when refactoring existing tests.
```
Comment (Hamms) [inline STYLEGUIDE.md]: nit:
```suggestion
However, it is important to acknowledge that much of our codebase (the majority of our codebase) does not actually follow this recommendation. This document serves as a guide to encourage moving towards these best practices when writing new code and or refactoring existing tests. 
```

File: STYLEGUIDE.md [hunk 6]
```diff
@@ -64,6 +64,79 @@ Fallback: https://github.com/styleguide/ruby
...
+
+When writing unit test we recommend to use [Specs Syntax](https://github.com/minitest/minitest?tab=readme-ov-file#specs-). Our default MiniTest Spec has been extended to include some RSpec features such as [`subject`](https://rspec.info/features/3-13/rspec-core/subject/explicit-subject/) or [`let!`](https://rspec.info/features/3-13/rspec-core/helper-methods/let/)
+
+This documentation represents a collection of references and resources for following best practices of Specs Syntax. It is intended to guide developers in writing clean, maintainable, and consistent unit tests using the Specs Syntax.
+
+However, it is important to acknowledge that much of our codebase (the majority of our codebase) does not actually follow this recommendation. This document serves as a guide to encourage moving towards these best practices in new code and when refactoring existing tests. 
+Additionally, note that multiple syntaxes can exist in the same test file, and this flexibility accommodates different styles and transitions.
+
+#### General Guidelines
+- Group-related tests using `describe` blocks.
```
Comment (Hamms) [inline STYLEGUIDE.md]: nit:
```suggestion
- Group related tests using `describe` blocks.
```

File: STYLEGUIDE.md (hunk 3 above)
Comment (nicklathe) [inline STYLEGUIDE.md:69]:
```suggestion
Our unit testing library is [Minitest](https://github.com/minitest/minitest). 
```

Comment (nicklathe) [inline STYLEGUIDE.md]:
```suggestion
Our unit testing library is [Minitest](https://github.com/minitest/minitest). 
```

File: STYLEGUIDE.md (hunk 4 above)
Comment (nicklathe) [inline STYLEGUIDE.md]:
```suggestion
When writing unit test we recommend using [Specs Syntax](https://github.com/minitest/minitest?tab=readme-ov-file#specs-). Our default MiniTest Spec has been extended to include some RSpec features such as [`subject`](https://rspec.info/features/3-13/rspec-core/subject/explicit-subject/) or [`let!`](https://rspec.info/features/3-13/rspec-core/helper-methods/let/)
```

File: STYLEGUIDE.md (hunk 5 above)
Comment (nicklathe) [inline STYLEGUIDE.md]:
```suggestion
However, it is important to acknowledge that most of our codebase does not actually follow this recommendation. This document serves as a guide to encourage moving towards these best practices in new code and when refactoring existing tests. 
```

File: STYLEGUIDE.md [hunk 7]
```diff
@@ -147,7 +220,7 @@ Default: https://google.github.io/styleguide/javascriptguide.xml
   });
   ```
 
-* <a name="js-operator-continuations"></a>
+- <a name="js-operator-continuations"></a>
```
Comment (nicklathe) [inline STYLEGUIDE.md]: Why the change from `*` to `-` here when the rest of this bulleted list uses `*`?

**Review by nicklathe — APPROVED** [review]
Some minor grammatical suggestions, nothing blocking.

File: STYLEGUIDE.md (hunk 7 above)
Comment (mgc1194) [reply STYLEGUIDE.md]: I was trying to make the formatting more consistent and changed all the `*`to `-`. But then I thought it wasn't needed in this PR and would take attention from the intended changes.

**Review by Hamms — COMMENTED** [review]
I'm personally skeptical of the value of recommending a new syntax over the patterns already established in our codebase (and I also find it significantly less readable than the syntax we currently use), but if we have broad buy-in from the engineering team as a whole that this is the direction we want to start heading in then I'm down to try to make it work.

In order to make it work, I think it's important that this proposal includes a plan to start converting existing tests to use the new syntax, including some idea of how we want to support existing functionality that's specific to the old syntax ([the `teardown_all` method](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/testing/setup_all_and_teardown_all.rb), for example). It's much more common for developers to author new code based on existing examples they can find in the codebase, than it is to begin by consulting the styleguide; with neither an enforcement mechanism nor a significant volume of examples in place, I worry that the disconnect between this documentation and what's actually in our codebase could be a source of confusion or even misinformation.

**Review by carl-codeorg — APPROVED** [review]

**Review by wilkie — APPROVED** [review]
LGTM. This brings us in line with accepted convention elsewhere and puts us in a better position to be making better use of our testing library and tools.

Would be nice to officially try to deprecate our existing tests and come up with a plan, now, to refactor them.

**Review by snickell — APPROVED** [review]
